File: data-jpa/src/main/java/top/bettercode/summer/data/jpa/query/mybatis/hibernate/MybatisLoader.java
package top.bettercode.summer.data.jpa.query.mybatis.hibernate;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.List;
import org.hibernate.HibernateException;
import org.hibernate.QueryException;
import org.hibernate.engine.spi.QueryParameters;
import org.hibernate.engine.spi.SessionFactoryImplementor;
import org.hibernate.engine.spi.SharedSessionContractImplementor;
import org.hibernate.loader.custom.CustomLoader;
import org.hibernate.loader.custom.CustomQuery;
import org.hibernate.loader.spi.AfterLoadAction;
import org.hibernate.transform.ResultTransformer;
import top.bettercode.summer.data.jpa.query.mybatis.MybatisResultTransformer;

/**
 * @author Peter Wu
 */
public class MybatisLoader extends CustomLoader {

  private boolean mybatisQuery;

  public MybatisLoader(CustomQuery customQuery, SessionFactoryImplementor factory) {
    super(customQuery, factory);
  }

  @Override
  protected List<?> getResultList(List results, ResultTransformer resultTransformer)
      throws QueryException {
    if (mybatisQuery) {
      return results;
    } else {
      return super.getResultList(results, resultTransformer);
    }
  }

  @Override
  public Object loadSingleRow(ResultSet resultSet, SharedSessionContractImplementor session,
      QueryParameters queryParameters, boolean returnProxies) throws HibernateException {
    if (mybatisQuery) {
      return ((MybatisResultTransformer) queryParameters.getResultTransformer()).transformResultSet(
          resultSet);
    } else {
      return super.loadSingleRow(resultSet, session, queryParameters, returnProxies);
    }
  }

  @Override
  protected List<?> processResultSet(ResultSet rs, QueryParameters queryParameters,
      SharedSessionContractImplementor session, boolean returnProxies,
      ResultTransformer forcedResultTransformer, int maxRows,
      List<AfterLoadAction> afterLoadActions) throws SQLException {
    if (mybatisQuery) {
      return ((MybatisResultTransformer) queryParameters.getResultTransformer()).transformListResultSet(rs,
          maxRows);
    } else {
      return super.processResultSet(rs, queryParameters, session, returnProxies,
          forcedResultTransformer, maxRows, afterLoadActions);
    }
  }

  @Override
  protected SqlStatementWrapper executeQueryStatement(QueryParameters queryParameters,
      boolean scroll, List<AfterLoadAction> afterLoadActions,
      SharedSessionContractImplementor session) throws SQLException {
    mybatisQuery = queryParameters.getResultTransformer() instanceof MybatisResultTransformer;
    return super.executeQueryStatement(queryParameters, scroll, afterLoadActions, session);
  }

  @Override
  protected void autoDiscoverTypes(ResultSet rs) {
    if (!mybatisQuery) {
      super.autoDiscoverTypes(rs);
    }
  }

}

File: data-jpa/src/main/kotlin/top/bettercode/summer/data/jpa/query/mybatis/hibernate/MybatisLoader.kt
package top.bettercode.summer.data.jpa.query.mybatis.hibernate

import org.hibernate.engine.spi.QueryParameters
import org.hibernate.engine.spi.SessionFactoryImplementor
import org.hibernate.engine.spi.SharedSessionContractImplementor
import org.hibernate.hql.internal.HolderInstantiator
import org.hibernate.loader.custom.CustomLoader
import org.hibernate.loader.custom.CustomQuery
import org.hibernate.loader.spi.AfterLoadAction
import org.hibernate.query.spi.ScrollableResultsImplementor
import org.hibernate.transform.ResultTransformer
import org.hibernate.type.Type
import top.bettercode.summer.data.jpa.query.mybatis.MybatisResultTransformer
import java.sql.ResultSet

/**
 * @author Peter Wu
 */
class MybatisLoader(customQuery: CustomQuery?, factory: SessionFactoryImplementor?) :
    CustomLoader(customQuery, factory) {
    private var mybatisQuery = false


    override fun scroll(
        queryParameters: QueryParameters?,
        returnTypes: Array<out Type>?,
        holderInstantiator: HolderInstantiator?,
        session: SharedSessionContractImplementor?
    ): ScrollableResultsImplementor {
        val resultTransformer = queryParameters?.resultTransformer
        if (resultTransformer is MybatisResultTransformer) {
            mybatisQuery = true
            resultTransformer.autoCloseResultSet = false
        }
        return super.scroll(
            queryParameters,
            returnTypes,
            if (mybatisQuery) HolderInstantiator.NOOP_INSTANTIATOR else holderInstantiator,
            session
        )
    }

    override fun getResultList(results: List<*>, resultTransformer: ResultTransformer?): List<*> {
        return if (mybatisQuery) {
            results
        } else {
            super.getResultList(results, resultTransformer)
        }
    }

    override fun loadSingleRow(
        resultSet: ResultSet, session: SharedSessionContractImplementor,
        queryParameters: QueryParameters, returnProxies: Boolean
    ): Any? {
        return if (mybatisQuery) {
            (queryParameters.resultTransformer as MybatisResultTransformer).transformResultSet(
                resultSet
            )
        } else {
            super.loadSingleRow(resultSet, session, queryParameters, returnProxies)
        }
    }

    override fun processResultSet(
        rs: ResultSet, queryParameters: QueryParameters,
        session: SharedSessionContractImplementor, returnProxies: Boolean,
        forcedResultTransformer: ResultTransformer?, maxRows: Int,
        afterLoadActions: List<AfterLoadAction>
    ): List<*> {
        return if (mybatisQuery) {
            (queryParameters.resultTransformer as MybatisResultTransformer).transformListResultSet(
                rs,
                maxRows
            )
        } else {
            super.processResultSet(
                rs, queryParameters, session, returnProxies,
                forcedResultTransformer, maxRows, afterLoadActions
            )
        }
    }

    override fun executeQueryStatement(
        queryParameters: QueryParameters,
        scroll: Boolean, afterLoadActions: List<AfterLoadAction>,
        session: SharedSessionContractImplementor
    ): SqlStatementWrapper {
        mybatisQuery = queryParameters.resultTransformer is MybatisResultTransformer
        return super.executeQueryStatement(queryParameters, scroll, afterLoadActions, session)
    }

    override fun autoDiscoverTypes(rs: ResultSet) {
        if (!mybatisQuery) {
            super.autoDiscoverTypes(rs)
        }
    }
}
